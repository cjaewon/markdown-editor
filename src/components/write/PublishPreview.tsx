import React from 'react';
import styled, { css } from 'styled-components';
import PublishSection from './PublishSection';
import palette from '../../lib/styles/palette';
import { ImageVector } from '../../static/svg';
import { ellipsis } from '../../lib/styles/utils';

const PublishPreviewBlock = styled(PublishSection)``;

const ThumbnailSizer = styled.div`
  width: 100%;
  padding-top: 55.11%;
  position: relative;
`;

const ThumbnailBlock = styled.div`
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  top: 0;
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.03);
`;

const MissingThumbnail = styled.div`
  background: ${palette.gray2};
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

const Image = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;

const UploadButton = styled.button`
  margin-top: 1rem;
  padding: 0.25rem 2rem;
  background: white;
  border-radius: 4px;
  box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.025);
  font-size: 1rem;
  line-height: 1.5;
  color: ${palette.teal5};
  outline: none;
  border: none;
  cursor: pointer;
  transition: all 0.125s ease-in;
  font-weight: bold;
  &:focus {
    box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.15);
  }
  &:hover {
    background: ${palette.gray0};
  }
`;

const PostInfo = styled.div`
  margin-top: 1.5rem;
  h4 {
    line-height: 1.5;
    margin: 0;
    display: block;
    font-size: 1.125rem;
    ${ellipsis};
  }
`;

const ShortDescriptionTextarea = styled.textarea`
  resize: none;
  width: 100%;
  border: none;
  outline: none;
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.03);
  line-height: 1.5;
  font-size: 0.875rem;
  height: 7.375rem;
  padding: 0.75rem 1rem;
  margin-top: 0.5rem;
`;

const TextLimit = styled.div<{ limit: boolean }>`
  text-align: right;
  margin-top: 0.25rem;
  font-size: 0.75rem;
  color: ${palette.gray6};
  ${props =>
    props.limit &&
    css`
      color: ${palette.red6};
    `}
`;

interface ThumbnailProps {
  thumbnail: string | null;
  onUploadClick: () => void;
}
const Thumbnail: React.FC<ThumbnailProps> = ({ onUploadClick, thumbnail }) => {
  return (
    <ThumbnailSizer>
      <ThumbnailBlock>
        {thumbnail ? (
          <Image src={thumbnail} data-testid="image" />
        ) : (
          <MissingThumbnail>
            <ImageVector />
            <UploadButton onClick={onUploadClick}>썸네일 업로드</UploadButton>
          </MissingThumbnail>
        )}
      </ThumbnailBlock>
    </ThumbnailSizer>
  );
};

const ThumbnailModifyBlock = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-bottom: 0.5rem;
  & > .actions {
    display: flex;
    align-items: center;
    button {
      background: none;
      outline: none;
      border: none;
      font-size: 1rem;
      color: ${palette.gray6};
      cursor: pointer;
      padding: 0;
      text-decoration: underline;
      &:hover {
        color: ${palette.gray5};
      }
      &:active {
        color: ${palette.gray7};
      }
    }
    .middledot {
      margin-left: 0.5rem;
      margin-right: 0.5rem;
      display: block;
      width: 2px;
      height: 2px;
      border-radius: 1px;
      background: ${palette.gray6};
    }
  }
`;
type ThumbnailModifyProps = {
  visible: boolean;
  onResetThumbnail: () => void;
  onChangeThumbnail: () => void;
};
const ThumbnailModify: React.FC<ThumbnailModifyProps> = ({
  visible,
  onResetThumbnail,
  onChangeThumbnail,
}) => {
  if (!visible) return null;
  return (
    <ThumbnailModifyBlock>
      <div className="actions">
        <button onClick={onChangeThumbnail}>재업로드</button>
        <div className="middledot" />
        <button onClick={onResetThumbnail}>제거</button>
      </div>
    </ThumbnailModifyBlock>
  );
};

const { useCallback } = React;

export interface PublishPreviewProps {
  title: string;
  description: string;
  thumbnail: string | null;
  onChangeDescription: (description: string) => void;
  onResetThumbnail: () => void;
  onUpload: () => any;
}

const PublishPreview: React.FC<PublishPreviewProps> = ({
  title,
  thumbnail,
  description,
  onChangeDescription,
  onUpload,
  onResetThumbnail,
}) => {
  // const descriptionToShow: string = description || defaultDescription;
  const onChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    onChangeDescription(e.target.value);
  };

  const onKeyDown = useCallback(
    (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
      if (e.keyCode === 13) {
        e.preventDefault();
      }
    },
    [],
  );
  return (
    <PublishPreviewBlock title="포스트 미리보기">
      <ThumbnailModify
        visible={!!thumbnail}
        onResetThumbnail={onResetThumbnail}
        onChangeThumbnail={onUpload}
      />
      <Thumbnail onUploadClick={onUpload} thumbnail={thumbnail} />
      <PostInfo>
        <h4>{title}</h4>
        <ShortDescriptionTextarea
          placeholder="당신의 포스트를 짧게 소개해보세요."
          value={description}
          onChange={onChange}
          onKeyDown={onKeyDown}
        />
        <TextLimit limit={description.length === 150}>
          {description.length}/150
        </TextLimit>
      </PostInfo>
    </PublishPreviewBlock>
  );
};

export default PublishPreview;
