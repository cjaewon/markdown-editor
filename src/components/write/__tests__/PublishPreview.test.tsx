import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PublishPreview, { PublishPreviewProps } from '../PublishPreview';

describe('PublishPreview', () => {
  const setup = (props: Partial<PublishPreviewProps> = {}) => {
    const initialProps: PublishPreviewProps = {
      title: '타이틀',
      description: '',
      onChangeDescription: () => {},
      onResetThumbnail: () => {},
      onUpload: () => {},
      thumbnail: null,
    };
    const utils = render(<PublishPreview {...initialProps} {...props} />);
    const textarea = utils.getByPlaceholderText(
      '당신의 포스트를 짧게 소개해보세요.',
    ) as HTMLTextAreaElement;
    return {
      textarea,
      ...utils,
    };
  };
  it('renders properly', () => {
    setup();
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  it('renders title', () => {
    const utils = setup({ title: '타이틀 입니다.' });
    utils.getByText('타이틀 입니다.');
  });

  it('calls onChangeDescription', () => {
    const onChangeDescription = jest.fn();
    const utils = setup({
      onChangeDescription,
    });
    fireEvent.change(utils.textarea, {
      target: {
        value: 'hello there',
      },
    });
    expect(onChangeDescription).toBeCalledWith('hello there');
  });
  it('shows the right length', () => {
    const utils = setup({ description: 'onetwothree' });
    utils.getByText('11/150');
  });
  it('text limit turns red', () => {
    const longDescription = new Array(150).fill('a').join('');
    const utils = setup({ description: longDescription });
    const textLimitDiv = utils.getByText('150/150');
    expect(window.getComputedStyle(textLimitDiv).color).toBe(
      'rgb(250, 82, 82)',
    );
  });
  it('calls onUpload', () => {
    const onUpload = jest.fn();
    const utils = setup({ onUpload });
    const uploadButton = utils.getByText('썸네일 업로드');
    fireEvent.click(uploadButton);
    expect(onUpload).toBeCalled();
  });
  it('shows a thumbnail', () => {
    const utils = setup({ thumbnail: 'https://images.velog.io/sample.png' });
    const image = utils.getByTestId('image') as HTMLImageElement;
    expect(image).toHaveAttribute('src', 'https://images.velog.io/sample.png');
    utils.getByText('재업로드');
    utils.getByText('제거');
  });
});
