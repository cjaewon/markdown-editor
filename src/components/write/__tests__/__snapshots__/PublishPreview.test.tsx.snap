// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublishPreview matches snapshot 1`] = `
<div>
  <section
    class="sc-bdVaJa ezdyV sc-bwzfXH uTxCW"
  >
    <h3>
      포스트 미리보기
    </h3>
    <div
      class="contents"
    >
      <div
        class="sc-htpNat fGfspC"
      >
        <div
          class="sc-bxivhb oVXVi"
        >
          <div
            class="sc-ifAKCX bnjEss"
          >
            <svg>
              vector-image.svg
            </svg>
            <button
              class="sc-bZQynM gnnAiV"
            >
              썸네일 업로드
            </button>
          </div>
        </div>
      </div>
      <div
        class="sc-gzVnrw fwQhVf"
      >
        <h4>
          타이틀
        </h4>
        <textarea
          class="sc-htoDjs cNItPr"
          placeholder="당신의 포스트를 짧게 소개해보세요."
        />
        <div
          class="sc-dnqmqq JcFnD"
        >
          0
          /150
        </div>
      </div>
    </div>
  </section>
</div>
`;
